<h1> Titanic Project--investigating the accuracy of various algorithims and decision trees with different features.
<h2> by Tessa de Vries
<h3> November 29, 2016

In [2]:
#code to import a csv and convert it to a numpy array
#import the numpy library
import numpy as np
#open the file
raw_data = open("titanic.csv")
# load the CSV file as a numpy matrix
dataset = np.loadtxt(raw_data, delimiter=",")
#'shape' is the dimensions of the matrix
print(dataset.shape)
# separate the data from the target attributes
X = dataset[:,0:3]
y = dataset[:,3]
print(X)
print(y)


(2201, 4)
[[ 1.  1.  1.]
 [ 1.  1.  1.]
 [ 1.  1.  1.]
 ..., 
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]]
[ 1.  1.  1. ...,  0.  0.  0.]


<h2> What is the average accuracy of the following three machine learning algorithms on this dataset: DecisionTree, KNeighbors, RandomForest?

In [3]:
from sklearn import datasets
import random
import statistics

from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .25)

from sklearn import tree
tree_classifier = tree.DecisionTreeClassifier();
tree_classifier = tree_classifier.fit(X_train, y_train)
tree_predictions = tree_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(tree_predictions, y_test))

0.791288566243


In [22]:
DecisionTree= [0.774954627949, 0.80943738657, 0.778584392015, 0.785843920145, 0.793103448276]
print("Mean accuracy of DecisionTree:")
print(statistics.mean(DecisionTree))


Mean accuracy of DecisionTree:
0.788384754991


In [5]:
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .25)

from sklearn import neighbors
knn_classifier = neighbors.KNeighborsClassifier()
knn_classifier = knn_classifier.fit(X_train, y_train)
knn_predictions = knn_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(knn_predictions, y_test))


0.789473684211


In [23]:
Neighbors= [0.651542649728, 0.787658802178, 0.773139745917, 0.802177858439, 0.771324863884]
print("Mean accuracy of Neighbors:")
print(statistics.mean(Neighbors))

Mean accuracy of Neighbors:
0.7571687840292


In [7]:
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .25)

from sklearn import ensemble
rf_classifier = ensemble.RandomForestClassifier()
rf_classifier = rf_classifier.fit(X_train, y_train)
rf_predictions = rf_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(rf_predictions, y_test))


0.793103448276


In [8]:
RandomForest= [0.776769509982, 0.787658802178, 0.764065335753, 0.785843920145, 0.747731397459]
print("Mean accuracy of RandomForest:")
print(statistics.mean(RandomForest))


Mean accuracy of RandomForest:
0.7724137931034


Based off the following code, the Decision Tree algorithm has the highest average accuracy. 

<h2> Using the most accurate algorithm, predict whether the following imaginary people would have survived:

a. An adult male in the 3rd class

b. An adult female in 1st class

c. A female child in 1st class

In [9]:
tree_predictions = tree_classifier.predict([[3, 1, 1]])
print(tree_predictions)

tree_predictions = tree_classifier.predict([[1, 1, 0]])
print(tree_predictions)

tree_predictions = tree_classifier.predict([[1, 0, 0]])
print(tree_predictions)



[ 0.]
[ 1.]
[ 1.]


An adult male in the 3rd class would die.
An adult female in 1st class would live.
A female child in 1st class would live. 

<h2> Which is the most important determining feature in this dataset?

In [10]:
import numpy as np
raw_data = open("titanic.csv")
dataset = np.loadtxt(raw_data, delimiter=",")
Xclass = dataset[:,0:1]
y = dataset[:,3]
from sklearn import datasets
import random
import statistics

from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(Xclass, y, test_size = .25)

from sklearn import tree
tree_classifier = tree.DecisionTreeClassifier();
tree_classifier = tree_classifier.fit(X_train, y_train)
tree_predictions = tree_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(tree_predictions, y_test))

0.713248638838


In [11]:
classs= [0.736842105263, 0.709618874773, 0.722323049002, 0.718693284936, 0.722323049002]
print("Accuracy of Decision Tree with just the Class Feature:")
print(statistics.mean(classs))


Accuracy of Decision Tree with just the Class Feature:
0.7219600725952


In [12]:
import numpy as np
raw_data = open("titanic.csv")
dataset = np.loadtxt(raw_data, delimiter=",")
Xage = dataset[:,1:2]
y = dataset[:,3]
from sklearn import datasets
import random
import statistics

from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(Xage, y, test_size = .25)

from sklearn import tree
tree_classifier = tree.DecisionTreeClassifier();
tree_classifier = tree_classifier.fit(X_train, y_train)
tree_predictions = tree_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(tree_predictions, y_test))



0.676950998185


In [13]:
age= [0.698729582577, 0.686025408348, 0.678765880218, 0.720508166969, 0.68058076225]
print("Accuracy of Decision Tree with just the Age Feature:")
print(statistics.mean(age))

Accuracy of Decision Tree with just the Age Feature:
0.6929219600724


In [14]:
import numpy as np
raw_data = open("titanic.csv")
dataset = np.loadtxt(raw_data, delimiter=",")
Xsex = dataset[:,2:3]
y = dataset[:,3]
from sklearn import datasets
import random
import statistics

from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(Xsex, y, test_size = .25)

from sklearn import tree
tree_classifier = tree.DecisionTreeClassifier();
tree_classifier = tree_classifier.fit(X_train, y_train)
tree_predictions = tree_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(tree_predictions, y_test))


0.773139745917


In [15]:
sex= [0.780399274047, 0.793103448276, 0.751361161525, 0.787658802178, 0.784029038113]
print("Accuracy of Decision Tree with just the Sex Feature:")
print(statistics.mean(sex))

Accuracy of Decision Tree with just the Sex Feature:
0.7793103448278


In [16]:
import numpy as np
raw_data = open("titanic.csv")
dataset = np.loadtxt(raw_data, delimiter=",")
Xminusclass = dataset[:,1:3]
y = dataset[:,3]
from sklearn import datasets
import random
import statistics

from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(Xminusclass, y, test_size = .25)

from sklearn import tree
tree_classifier = tree.DecisionTreeClassifier();
tree_classifier = tree_classifier.fit(X_train, y_train)
tree_predictions = tree_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(tree_predictions, y_test))

0.758620689655


In [17]:
minusclass= [0.762250453721, 0.767695099819, 0.764065335753, 0.767695099819, 0.773139745917 ]
print("Accuracy of Decision Tree without the Class Feature:")
print(statistics.mean(minusclass))

Accuracy of Decision Tree without the Class Feature:
0.7669691470058


In [18]:
import numpy as np
raw_data = open("titanic2.csv")
dataset = np.loadtxt(raw_data, delimiter=",")
Xminusage = dataset[:,1:3]
y = dataset[:,3]
from sklearn import datasets
import random
import statistics

from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(Xminusage, y, test_size = .25)

from sklearn import tree
tree_classifier = tree.DecisionTreeClassifier();
tree_classifier = tree_classifier.fit(X_train, y_train)
tree_predictions = tree_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(tree_predictions, y_test))

0.764065335753


In [19]:
minusage= [0.778584392015, 0.785843920145, 0.767695099819, 0.796733212341, 0.794918330309 ]
print("Accuracy of Decision Tree without the Age Feature:")
print(statistics.mean(minusage))

Accuracy of Decision Tree without the Age Feature:
0.7847549909258


In [20]:
import numpy as np
raw_data = open("titanic3.csv")
dataset = np.loadtxt(raw_data, delimiter=",")
Xminussex = dataset[:,1:3]
y = dataset[:,3]
from sklearn import datasets
import random
import statistics

from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(Xminussex, y, test_size = .25)

from sklearn import tree
tree_classifier = tree.DecisionTreeClassifier();
tree_classifier = tree_classifier.fit(X_train, y_train)
tree_predictions = tree_classifier.predict(X_test)

from sklearn.metrics import accuracy_score
print(accuracy_score(tree_predictions, y_test))

0.691470054446


In [21]:
minussex= [0.740471869328, 0.733212341198, 0.736842105263, 0.705989110708, 0.713248638838]
print("Accuracy of Decision Tree without the Sex Feature:")
print(statistics.mean(minussex))

Accuracy of Decision Tree without the Sex Feature:
0.725952813067


When removing all features besides the single targeted feature (either class, age, or sex), the single feature that produced the highest Decision Tree accuracy was the sex feature at an accuracy of 77.93%. However, the other single feature based Decision Trees produced close accuracies at 72.196% (class) and 69.29% (age). 

When removing just one feature from the Decision Tree (tree was either class and age, sex and age, or class and sex), the tree that produced the highest Decision Tree accuracy was the tree that disregarded the age feature with an accuracy of 78.475%. However, the other trees had similar accuracy at 72.595% (disregarding sex feature) and 76.697% (disregarding class feature). 

The average accuracy of the Decision Tree when considering all features was 78.838%. That being said, based off the previous calculations, the decision tree is fairly just as accurate when removing the age feature. And, even more shocking, the decision tree with just the sex feature had a very similar accuracy with a difference of less than 1%. On top of that, the tree without the sex feature had the lowest average accuracy at 72.595%. Therefore, one may conclude that the sex feature is the most determining feature.

<h2> Write a conclusion - what did you learn overall?  What are you still wondering about?

Overall I learned that sex was a very important determining characteristic for life or death in the Titanic. This was somewhat a suprise for me as I thought that age would be just as significant in determining life or death. I thought this because in the movie all the children were saved by the rescue boats first, but then again women were saved alongside the children so the sex feature being most important isn't that suprising. 

I wonder what percentange of all passengers lived. I also still wonder within each feature which percentage lived. For example, in the sex feature I wonder what percent of men lived versus what percent of women lived. And for the age feature, I wonder which percent of adults lived versus what percent of children lived. 